v-next/hardhat-errors/src/errors.ts

```diff
@@ -130,68 +186,27 @@ function getErrorCode(errorDescriptor: ErrorDescriptor): string {
  */
 export function applyErrorMessageTemplate(
   template: string,
-  values: ErrorMessageTemplateArguments,
+  values: Record<string, ErrorMessageTemplateValue>,
 ): string {
-  return _applyErrorMessageTemplate(template, values, false);
-}
-
-function _applyErrorMessageTemplate(
-  template: string,
-  values: ErrorMessageTemplateArguments,
-  isRecursiveCall: boolean,
-): string {
-  if (!isRecursiveCall) {
-    for (const variableName of Object.keys(values)) {
-      assertHardhatInvariant(
-        /^[a-zA-Z][a-zA-Z0-9]*$/.test(variableName),
-        `Trying to apply error template but variable "${variableName}" is invalid. Variable names can only include ascii letters and numbers, and start with a letter.`,
-      );
-
-      const variableTag = `%${variableName}%`;
-
-      assertHardhatInvariant(
-        template.includes(variableTag),
-        `Trying to apply error template but variable "${variableName}" is not present in the template.`,
-      );
-    }
-  }
-
-  if (template.includes("%%")) {
-    return template
-      .split("%%")
-      .map((part) => _applyErrorMessageTemplate(part, values, true))
-      .join("%");
-  }
-
-  for (const variableName of Object.keys(values)) {
-    let value: string;
-
+  return template.replaceAll(/{(.*?)}/g, (_match, variableName) => {
     const rawValue = values[variableName];
 
     if (rawValue === undefined) {
-      value = "undefined";
-    } else if (rawValue === null) {
-      value = "null";
-    } else if (typeof rawValue === "bigint") {
-      value = `${rawValue}n`;
-    } else if (
-      typeof rawValue === "object" &&
-      !rawValue.hasOwnProperty("toString")
-    ) {
-      value = JSON.stringify(rawValue);
-    } else {
-      value = rawValue.toString();
+      return "undefined";
     }
 
-    const variableTag = `%${variableName}%`;
+    if (rawValue === null) {
+      return "null";
+    }
 
-    assertHardhatInvariant(
-      !/%([a-zA-Z][a-zA-Z0-9]*)?%/.test(value),
-      `Trying to apply an error template but variable "${variableName}" has a value that contains a variable tag.`,
-    );
+    if (typeof rawValue === "bigint") {
+      return `${rawValue}n`;
+    }
 
-    template = template.replaceAll(variableTag, value);
-  }
+    if (Array.isArray(rawValue)) {
```

Comment:
@schaable when merging this conflict I realized that the `hasOwnProperty` method didn't allow for classes to define their `toString`, so I switched it to explicitly handling arrays.